workshop/4 - Reliability/README.md

```diff
@@ -0,0 +1,208 @@
+
+    ![Screenshot of circuit breaker exception](../images/4-Reliability/circuit-breaker.png)
+
+
+### Checking the Main Relecloud Application to check Azure service SDKs and client libraries
+
+Most Azure services and client SDKs have a built-in retry mechanism. You should use the built-in retry mechanism for Azure services to expedite the implementation. Let's see how the retry pattern is implemented in the main sample using Entity Framework Core's built-in retry mechanism.
+
+1. Select the main **Relecloud** solution.
+1. Open the **Relecloud.CallCenter.Api** project's **Startup.cs** file.
+1. Browse to the `AddConcertContextServices` method. This method configures the Entity Framework Core context for the application.
+1. Look at the `services.AddDbContextPool<ConcertDataContext>` call on the following code:
+
+    ```csharp
+    services.AddDbContextPool<ConcertDataContext>(options => options.UseSqlServer(sqlDatabaseConnectionString,
+        sqlServerOptionsAction: sqlOptions =>
+        {
+            sqlOptions.EnableRetryOnFailure(
+            maxRetryCount: 5,
+            maxRetryDelay: TimeSpan.FromSeconds(3),
+            errorNumbersToAdd: null);
+        }));
+    ```
+
+    The `EnableRetryOnFailure` method enables the built-in retry mechanism for Entity Framework Core. The default retry policy is to retry up to 6 times with a 2-second delay between retries. You can change the default retry policy by passing a `MaxRetryCount` and `MaxRetryDelay` to the `EnableRetryOnFailure` method.
+
+Now we'll use the Azure App Configuration's SDK to add a retry policy to communication with the Azure App Configuration service.
+
+
+1. Open the **Program.cs** file from the same project.
+1. Browse to the `builder.Configuration.AddAzureAppConfiguration` call. It should be on line 11.
+1. Inspect the following code onto the end of the `ConfigureKeyVault` method.
+
+    ```csharp
+    .ConfigureClientOptions(options =>
+    {
+        options.Retry.MaxRetries = 5;
+        options.Retry.MaxDelay = TimeSpan.FromSeconds(3);
+    });
+    ```
+
+    The `ConfigureClientOptions` method configures the retry policy for the Azure App Configuration SDK. The default retry policy is to retry up to 3 times with a 1-second delay between retries. You can change the default retry policy by passing a `MaxRetries` and `MaxDelay` to the `ConfigureClientOptions` method.
+
+    The entire Azure App Configuration configuration should look like this:
+
+    ```csharp
+    builder.Configuration.AddAzureAppConfiguration(options =>
+    {
+        options
+            .Connect(new Uri(builder.Configuration["Api:AppConfig:Uri"]), new DefaultAzureCredential())
+            .ConfigureKeyVault(kv =>
+            {
+                // Some of the values coming from Azure App Configuration are stored Key Vault, use
+                // the managed identity of this host for the authentication.
+                kv.SetCredential(new DefaultAzureCredential());
```

Comment:
You should be able to use a single azure credential for the lifetime of the application. See the updated pattern in the repo for creating one and passing it through